bindings/c/src/c_io_adapter.cc

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "c_io_adapter.h"
+
+#include "vw/experimental/types.h"
+
+#include "vw_exception.h"
+
+c_reader::c_reader(void* context, VWReadFunc* read)
+    : VW::io::reader(false /* is_resettable */), _context(context), _read_func(read)
+{
+}
+
+ssize_t c_reader::read(char* buffer, size_t num_bytes)
+{
+  if (_read_func == nullptr) { THROW("Read function is not set. Should not be possible.") }
+
+  size_t bytes_read = 0;
+  auto result = _read_func(_context, buffer, num_bytes, &bytes_read);
```

Comment:
what is `context`? Should we also check if it is `nullptr`?

Reply:
It's an opaque pointer passed to the user provided function, it allows it to have state. It's valid for it to be a nullptr